### ISSUE TITLE
List of dependencies to upgrade

### ISSUE THREAD

**Issue by IamLizu** [expressjs/discussions#256]
As part of the plan of publishing `5.0`, we need to upgrade the dependencies. The idea is to create better visibility on these deps. Please note that the list has been compared against [5.0 package.json](https://github.com/expressjs/express/blob/5.x/package.json).

Dependencies that are not directly owned by [expressjs](https://github.com/expressjs) / [PillarsJS](https://github.com/pillarsjs) / [jshttp](https://github.com/jshttp).

### List

- [x] ~~array-flatten~~ (removed at https://github.com/expressjs/express/pull/5677)
- [x] cookie-signature - used 1.0.6 - current 1.2.1 (maybe fixed by https://github.com/expressjs/express/pull/5833)
- [x] debug - used 3.1.0 - current 4.3.6 (fixed by https://github.com/expressjs/express/pull/5829)
- [x] depd
- [x] encodeurl
- [x] escape-html
- [x] etag
- [x] fresh
- [x] merge-descriptors - used 1.0.1 - current 2.0.0 (fixed by https://github.com/expressjs/express/pull/5782)
- [x] methods
- [x] path-is-absolute (deprecated, removed by https://github.com/expressjs/express/pull/5830) 
- [x] qs - used 6.11.0 - current 6.13.0 (fixed by https://github.com/expressjs/express/pull/5847)
- [x] safe-buffer
- [x] setprototypeof
- [x] utils-merge

#### Reference

- #233

**Comment by carpasse** [comment]
I reviewed all the dependencies, and below is the list of external dependencies that need to be updated:

| Dependency                                              | Current Version | Updated Version | Status/Notes                                          | License                                                                                                                                             |
|---------------------------------------------------------|-----------------|-----------------|------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------|
| [`cookie-signature`](https://www.npmjs.com/package/cookie-signature)  | 1.0.6           | 1.2.1           | Update available                                     | [MIT](https://en.wikipedia.org/wiki/MIT_License)                                                                                                     |
| [`debug`](https://www.npmjs.com/package/debug)                         | 3.1.0           | 4.3.6           | Update available                                     | [MIT](https://en.wikipedia.org/wiki/MIT_License)                                                                                                     |
| [`depd`](https://www.npmjs.com/package/depd)                           | 2.0.0           | -               | Already latest (6 years without update)              | [MIT](https://en.wikipedia.org/wiki/MIT_License)                                                                                                     |
| [`escape-html`](https://www.npmjs.com/package/escape-html)             | ~1.0.3          | -               | Already latest (9 years without update)              | [MIT](https://en.wikipedia.org/wiki/MIT_License)                                                                                                     |
| [`merge-descriptors`](https://www.npmjs.com/package/merge-descriptors) | 1.0.1           | 2.0.0           | Update available                                     | [MIT](https://en.wikipedia.org/wiki/MIT_License)                                                                                                     |
| [`once`](https://www.npmjs.com/package/once)                           | 1.4.0           | -               | Already latest (8 years without update)              | [ISC](https://en.wikipedia.org/wiki/ISC_license)                                                                                                     |
| [`path-is-absolute`](https://www.npmjs.com/package/path-is-absolute)   | 1.0.1           | -               | Deprecated/unmaintained                              | [N/A](https://www.npmjs.com/package/path-is-absolute)                                                                                                |
| [`qs`](https://www.npmjs.com/package/qs)                               | 6.11.0          | 6.13.0          | Update available                                     | [BSD-3-Clause](https://en.wikipedia.org/wiki/BSD_licenses#3-clause)                                                                                  |
| [`safe-buffer`](https://www.npmjs.com/package/safe-buffer)             | 5.2.1           | -               | Already latest (5 years without update)              | [MIT](https://en.wikipedia.org/wiki/MIT_License)                                                                                                     |
| [`setprototypeof`](https://www.npmjs.com/package/setprototypeof)       | 1.2.0           | -               | Already latest (5 years without update)              | [ISC](https://en.wikipedia.org/wiki/ISC_license)                                                                                                     |
| [`utils-merge`](https://www.npmjs.com/package/utils-merge)             | 1.0.1           | -               | Already latest (7 years without update)              | [MIT](https://en.wikipedia.org/wiki/MIT_License)                                                                                                     |

**Comment by IamLizu** [comment]
@carpasse I think my initial comment unclear, by checked items, I meant those do not need intervention. Only the unchecked items need to be taken care of.

In any case, thank you for putting in the effort.

And it appears you and I agree on the list. Perhaps we can now just check the ones that are upgraded, once their respective PR is merged?

**Comment by wesleytodd** [comment]
Hey! I was working through these as well (until work and life and security stuff all started happening at once) and I was attempting to track progress in #233. Would it be a good thing to add these last items to that todo list instead of in a separate issue?

**Comment by IamLizu** [comment]
Hey @wesleytodd 👋 

Yes, perhaps it would be better to keep the track there. 

I am unable to edit the last push discussion though.

**Comment by wesleytodd** [comment]
Could you just comment them in there for now?

**Comment by wesleytodd** [comment]
For that lib, you should check out this issue: https://github.com/tj/node-cookie-signature/issues/36#issuecomment-2253673707

**Comment by carpasse** [comment]
@IamLizu I am sorry I missed the message and started working in cookie signature when I connected this morning. I've closed the PR please ignore it.

I would like to work on upgrading `debug` dependency from `3.1.0` to `4.3.6`

cc @wesleytodd @UlisesGascon @IamLizu

**Comment by carpasse** [comment]
@wesleytodd @UlisesGascon [Express 5.0](https://github.com/expressjs/express/blob/5.0) is still [using `path-is-absolute`](https://github.com/expressjs/express/blob/5.0/package.json#L51) dependency which is [deprecated](https://www.npmjs.com/package/path-is-absolute). I would like to remove it since it will no longer be necessary for 5.0

**Comment by carpasse** [comment]
@wesleytodd @UlisesGascon @IamLizu I would like to work on updating [qs](https://www.npmjs.com/package/qs) dep from 6.11.0 to 6.13.0

**Comment by IamLizu** [comment]
Hey @carpasse 👋 

Alright, and for visibility, I am updating the main comment tagging your PRs.

**Comment by carpasse** [comment]
Question, what branch should we base the PRs to update the dependencies on?
branch [5.0](https://github.com/expressjs/express/tree/5.0) or branch [5-merge](https://github.com/expressjs/express/tree/5-merge)

**Comment by wesleytodd** [comment]
> [Express 5.0](https://github.com/expressjs/express/blob/5.0) is still [using path-is-absolute](https://github.com/expressjs/express/blob/5.0/package.json#L51) dependency which is [deprecated](https://www.npmjs.com/package/path-is-absolute). I would like to remove it since it will no longer be necessary for 5.0

Sounds like a good thing to remove. 👍 

> I would like to work on updating [qs](https://www.npmjs.com/package/qs) dep from 6.11.0 to 6.13.0

I think this has already been done somewhere. Let me look, but IIRC that release was an ask of ours. I will ping in slack about it.

> Question, what branch should we base the PRs to update the dependencies on?
branch [5.0](https://github.com/expressjs/express/tree/5.0) or branch [5-merge](https://github.com/expressjs/express/tree/5-merge)

`5.0` is the correct branch. That one is Chris working on merging in some changes from `master` which did not merge cleanly.

**Comment by IamLizu** [comment]
Hi @wesleytodd 👋 

> I think this has already been done somewhere. Let me look, but IIRC that release was an ask of ours. I will ping in slack about it.

I tried to search in the PRs but couldn't find qs being upgraded in anyone. I think I must have missed  something.

**Comment by carpasse** [comment]
@wesleytodd I had a chat with @UlisesGascon and the `qs` dep was updated on `body-parser` but not on `express` therefore I've created the PR.

**Comment by IamLizu** [comment]
Awesome!

Now that we have a PR against each of the pending deps, can we fast track these PRs and get them to land?

cc: @wesleytodd @UlisesGascon

**Comment by wesleytodd** [comment]
Everything above is checked off. I still have a few in #233 with remaining changes to land, but does this mean we are able to close this one?

**Comment by IamLizu** [comment]
Yes, we can close this 🎉 

I believe this issue has served its purpose on tracking the upgrades of dependencies of `express` which are not owned by `express`.   I hope it also makes the last point of "Pending things (Express):" in #233 checkable as well.

Thank you to everyone involved in upgrading these deps.

**Comment by bjohansebas** [comment]
hey @IamLizu `encodeurl` is outdated. The latest version is 2.0.0, while the one in the package is 1.0.2. It would be good to update it."

**Comment by IamLizu** [comment]
@bjohansebas `encodeurl` is owned by PillarJS. We were only focusing on the ones that not owned by Expressjs / PillarJS / JSHttp. Since it was already mentioned in the first comment, I didn't mention it explicitly in my last comment. 

I hope its clear now 🎉
